proxy/src/proxy.rs

```diff
@@ -209,9 +209,10 @@ async fn handshake<S: AsyncRead + AsyncWrite + Unpin>(
                     if let Some(tls) = tls.take() {
                         // Upgrade raw stream into a secure TLS-backed stream.
                         // NOTE: We've consumed `tls`; this fact will be used later.
-                        stream = PqStream::new(
-                            stream.into_inner().upgrade(tls.to_server_config()).await?,
-                        );
+                        // Client shouldn't send any data before TLS handshake,
+                        // so ignore pqstream read buffer.
+                        let (raw, _) = stream.into_inner();
+                        stream = PqStream::new(raw.upgrade(tls.to_server_config()).await?);
```

Comment:
It seems error-prone to ignore the pqstream read buffer. If we expect it to be empty, we should check that it's empty and throw an error and close the connection if it's not.

But couldn't the client pipeline the TLS ClientHello message and the Postgres SSLRequest message? I though the serverless driver actually does that, to avoid one round trip, or at least there was some discussion on doing that.

Reply:
> If we expect it to be empty, we should check that it's empty and throw an error and close the connection if it's not.

Makes sense, will fix.

> But couldn't the client pipeline the TLS ClientHello message and the Postgres SSLRequest message? 

Hm, in theory it can. In practice, in case of serverless driver TLS is established outside (in http, so websocket connection is already secured), so postgres connection is plaintext.

Reply:
Or I misunderstand this?

Reply:
Poking serverless driver, I think I am not. However, driver talks about experiment encryption on PG level, so this should be supported. I'll consider chaining streams.

Reply:
> > But couldn't the client pipeline the TLS ClientHello message and the Postgres SSLRequest message?
> 
> Hm, in theory it can. In practice, in case of serverless driver TLS is established outside (in http, so websocket connection is already secured), so postgres connection is plaintext.

Yeah, it's a bit weird to switch to TLS at this level, when the websockets connection is already TLS protected. You get TLS within TLS.

Perhaps we should disallow it altogether? Then again, there are reasons to do it:
- if the websockets proxy lives somewhere else, then you might indeed want to have another TLS layer to protect the connection from the websockets proxy to the database. That situation isn't relevant for our deployment, but if someone runs their own websockets proxy, it might be.
-  if you e.g. have a libpq client or application that requires TLS, and doesn't know that it's already protected by the outer websockets TLS layer.
-  SCRAM channel binding might be another reason. The libpq authentication handshake needs to know about the TLS connection for that
- If the connection to the websockets proxy is unencrypted, i.e "ws://" rather than "wss://".

Reply:
see https://github.com/neondatabase/neon/pull/3821#issuecomment-1469804954